### PR TITLE
static to non-static conversion for methods in Index[Merger/Maker/IO]

### PR DESCRIPTION
**Pull request by himanshug**
so that appropriate dependencies can be injected specially ObjectMapper that can handle extensions as well without each having their own static Injector creation.

**Release Notes update:**
this PR removes the support for deprecated property `druid.processing.bitmap` and makes Concise be default.

### CODE REVIEW

**Comment by xvrl** [comment]
@himanshug nice! I've been waiting so long for that day. @cheddar you owe @himanshug a beer :)

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java [hunk 1]
```diff
@@ -84,6 +87,9 @@
   public static final Splitter tabSplitter = Splitter.on("\t");
   public static final Joiner tabJoiner = Joiner.on("\t");
   public static final ObjectMapper jsonMapper;
```
Comment (xvrl) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: should we also update those constants above to uppercase to be consistent while we're at it?

Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: sure, will do that

File: indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java [hunk 2]
```diff
@@ -107,12 +112,24 @@ private static String makeDatasource(FireDepartment fireDepartment)
   @JsonIgnore
   private volatile QueryRunnerFactoryConglomerate queryRunnerFactoryConglomerate = null;
 
+  @JsonIgnore
+  private volatile IndexMerger indexMerger;
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java]: why volatile and not final?

File: indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTest.java [hunk 3]
```diff
@@ -102,6 +104,14 @@
 @RunWith(Parameterized.class)
 public class IngestSegmentFirehoseFactoryTest
 {
+  private static final IndexMerger INDEX_MERGER;
+  private static final IndexIO INDEX_IO;
+
+  static {
```
Comment (drcrallen) [inline indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTest.java]: can these be moved to a `@BeforeClass`?

File: indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTimelineTest.java [hunk 4]
```diff
@@ -99,11 +101,20 @@
...
 
+  private static final IndexMerger INDEX_MERGER;
+  private static final IndexIO INDEX_IO;
+
   private final IngestSegmentFirehoseFactory factory;
   private final File tmpDir;
   private final int expectedCount;
   private final long expectedSum;
 
+  static {
```
Comment (drcrallen) [inline indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTimelineTest.java]: same question about `@BeforeClass`

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 5]
```diff
@@ -113,64 +106,116 @@
...
-                // this property is deprecated, use IndexSpec instead
-                JsonConfigProvider.bind(binder, "druid.processing.bitmap", BitmapSerdeFactory.class);
-              }
-            }
-        )
-    );
-    mapper = injector.getInstance(ObjectMapper.class);
-    columnConfig = injector.getInstance(ColumnConfig.class);
-    CONFIGURED_BITMAP_SERDE_FACTORY = injector.getInstance(BitmapSerdeFactory.class);
+  public void validateTwoSegments(File dir1, File dir2) throws IOException
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: can you move this to its prior location in file to minimize diff crazyness?

**Comment by drcrallen** [comment]
This is cool.

Will probably want to callout in release notes because this might break extensions (it will break some extensions I'm tinkering with).

File: indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java (hunk 2 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java]: just to be consistent with variables above, will make both final

Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java]: it appears you can't have final and volatile both, I'm keeping it volatile like other variables just to be consistent for now.

Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java]: It will screw things up if someone tries to change it, right?

Comment (xvrl) [inline indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java]: @himanshu volatile doesn't make sense for a final variable, since it never changes.

Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java]: actually nevermind, it looks like if someone were to change the value it would get used as expected

File: indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTest.java (hunk 3 above)
Comment (himanshug) [inline indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTest.java]: with @BeforeClass I can't keep the variables final, so using static.

File: indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTimelineTest.java (hunk 4 above)
Comment (himanshug) [inline indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTimelineTest.java]: with @BeforeClass I can't keep the variables final, so using static.

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 5 above)
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: this uses `loadIndex` in IndexIO which is not static anymore. if I keep it inside prior location, DefaultIndexIOHandler then that class can not be declared static. It is simpler to move those.

**Comment by himanshug** [comment]
@drcrallen sure, we can call this out in release notes. I am putting a note on PR description too.

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java (hunk 1 above)
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: done

File: indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java (hunk 2 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java]: I will do another PR to convert them from volatile -> final with UTs to ensure that nothing is broken by that change.

Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java]: s/volatile/final where possible in this class, there are existing serde tests for this so change is safe.

**Comment by himanshug** [comment]
@xvrl @drcrallen all comments have been taken care of, pls see as this PR is required to unblock  #1728

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java [hunk 6]
```diff
@@ -107,8 +113,11 @@ public void configure(Binder binder)
             new IndexingHadoopModule()
         )
     );
-    jsonMapper = injector.getInstance(ObjectMapper.class);
+    JSON_MAPPER = injector.getInstance(ObjectMapper.class);
     properties = injector.getInstance(Properties.class);
```
Comment (gianm) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: I think `properties` can be removed- it was only used for the "hack to get druid.processing.bitmap property passed down" which you've now got rid of

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 7]
```diff
@@ -369,7 +419,13 @@ public static void validateRowValues(
 
   public static class DefaultIndexIOHandler implements IndexIOHandler
   {
-    private static final Logger log = new Logger(DefaultIndexIOHandler.class);
+    private final Logger log = new Logger(DefaultIndexIOHandler.class);
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: The logger should still be static

File: processing/src/main/java/io/druid/segment/IndexSpec.java [hunk 8]
```diff
@@ -98,7 +99,7 @@ public IndexSpec(
     Preconditions.checkArgument(metricCompression == null || COMPRESSION_NAMES.contains(metricCompression),
                                 "Unknown compression type[%s]", metricCompression);
 
-    this.bitmapSerdeFactory = bitmapSerdeFactory != null ? bitmapSerdeFactory : IndexIO.CONFIGURED_BITMAP_SERDE_FACTORY;
+    this.bitmapSerdeFactory = bitmapSerdeFactory != null ? bitmapSerdeFactory : new ConciseBitmapSerdeFactory();
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexSpec.java]: Does this mean `druid.processing.bitmap.type` will no longer work to set the default bitmap type?

Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexSpec.java]: ah, yes, it does, and you mentioned it in the PR summary.

Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexSpec.java]: The property is in the docs now, and not marked deprecated, so maybe it's worth keeping it but marking it as deprecated now. Or if we are ok with removing the config now, then it should also be removed from the docs.

**Comment by gianm** [comment]
this PR looks good to me after figuring out what we want to do with `druid.processing.bitmap.type` - either remove it from the docs, or add support back in but mark it deprecated in the docs.

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java (hunk 6 above)
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: removed

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 7 above)
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: done

File: processing/src/main/java/io/druid/segment/IndexSpec.java (hunk 8 above)
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/IndexSpec.java]: removed the prop from docs as well.

**Comment by himanshug** [comment]
@gian support for `druid.processing.bitmap.type` becomes complicated inside IndexSpec no arg constructor once we lose static properties object inside IndexIO  also it needs hacks to support it inside hadoop indexing as well . Code said it was deprecated (dint look at the doc at the time) and hence removed it to keep things simple. Since this is going inside a new major release, I believe it is OK.

**Comment by gianm** [comment]
@himanshug ok sounds fair

**Comment by gianm** [comment]
:+1: looks good to me

**Comment by drcrallen** [comment]
:+1:

File: indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java [hunk 9]
```diff
@@ -151,7 +155,9 @@ public IndexTask(
       @JsonProperty("resource") TaskResource taskResource,
       @JsonProperty("spec") IndexIngestionSpec ingestionSchema,
       @JacksonInject ObjectMapper jsonMapper,
-      @JsonProperty("context") Map<String, Object> context
+      @JsonProperty("context") Map<String, Object> context,
+      @JacksonInject IndexMerger indexMerger,
+      @JacksonInject IndexIO indexIO
```
Comment (xvrl) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: Injecting IndexIO and IndexMerger might be a problem for external libraries relying on those classes purely for serialization and de-serialization, such as tranquility. I wonder if it would be possible to separate the parts necessary for serde from the parts needed to run the task.

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: Could put them in the TaskToolbox instead.

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java [hunk 10]
```diff
@@ -107,8 +109,10 @@ public void configure(Binder binder)
             new IndexingHadoopModule()
         )
     );
-    jsonMapper = injector.getInstance(ObjectMapper.class);
-    properties = injector.getInstance(Properties.class);
+    JSON_MAPPER = injector.getInstance(ObjectMapper.class);
+    INDEX_IO = injector.getInstance(IndexIO.class);
+    INDEX_MERGER = injector.getInstance(IndexMerger.class);
+    INDEX_MAKER = injector.getInstance(IndexMaker.class);
```
Comment (xvrl) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: I wonder if it would be possible to make those fields non-static as well? This might make it easier to have configurable defaults, even for hadoop tasks.

File: indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java (hunk 9 above)
Comment (xvrl) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: the same concern also applies to all the other tasks, and classes that, until now, did not rely on jackson injection.

**Comment by xvrl** [comment]
@himanshug I had marked default bitmap config deprecated mainly because it is possible to specify it in the index spec ever since we added compression, whereas before that property was the only way to specify it. It might still be nice to be able to specify a default, but I'm also not wedded to it if you think it's too cumbersome to maintain. A more general way to specify indexing default might be nice in the future though.

File: indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java (hunk 9 above)
Comment (xvrl) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: we should also add some ser-de tests for tasks, I don't think we have any.

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: there's a lot in the TaskSerdeTest class

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: Fwiw I think tranquility will probably be able to handle these, as long as its injector has the ability to make these classes (which if it doesn't, could be added). Putting them in the TaskToolbox is cool too if we want to go that way.

Comment (xvrl) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: @gianm my bad, missed the task serde test. In general I think it would be nice to move away from jackson injecting if we ever want to make those classes available in a proper API or client library.

**Comment by xvrl** [comment]
other than my concerns about jackson injection, the rest looks fine to me.

File: indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java (hunk 9 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: @xvrl @gian if we put it inside TaskToolbox then TaskToolbox needs to be injected , I don't think we can avoid injection.
libraries will probably need to setup dummy objects for injection for all these.

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java (hunk 10 above)
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: nice idea, but, I would do that in a separate PR and not hold this one. :)

Comment (xvrl) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: yes, doesn't have to block this

File: indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java (hunk 9 above)
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: @himanshug the toolbox shouldn't need to be injected just for serialization, as it's only needed if you're calling `run`. That wouldn't happen if the *Task class is just being used by something that wants to generate a request to submit to the overlord (like tranquility does). Right?

I am ok with doing it either way though. Tranquility already has to set up an ObjectMapper with a bunch of Druid modules (for stuff like AggregatorFactories and Firehoses) and so needing this other stuff too wouldn't be much worse.

Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: @xvrl @gianm ok removed them from individual task constructors, added to TaskToolbox instead and using from same.
hopefully this will reduce some burden on tranquility side.

**Comment by himanshug** [comment]
@xvrl all comments addressed

**Comment by gianm** [comment]
changes looks good, :+1:

**Comment by xvrl** [comment]
:+1:
